From ca93cda4b820a3f7d539b0207c0ddce2171b3082 Mon Sep 17 00:00:00 2001
From: Joerg Bornemann <joerg.bornemann@qt.io>
Date: Tue, 25 Apr 2023 10:14:54 +0200
Subject: [PATCH 3/4] CMake: Fix qt5_create_translation for same-named .ts
 files

Consider a project where .ts files are structured by directory instead
of by file name:

translations/chinese/myapp.ts
translations/czech/myapp.ts
translations/french/myapp.ts
translations/german/myapp.ts
translations/italian/myapp.ts

In that case, qt5_create_translations failed at configure time with the
following error:

  Attempt to add a custom rule to output

    /build/myapp/.lupdate/myapp.ts.stamp.rule

  which already has a custom rule.

Fix that by constructing a .ts.stamp file path that takes the relative
path from CMAKE_CURRENT_SOURCE_DIR to the .ts file into account.

Additionally, pass the paths of the timestamp files to
qt5_add_translation with the __QT_INTERNAL_TIMESTAMP_FILES multi-value
argument to avoid having to recalculate the paths of the timestamp
files.

Pick-to: 6.5 5.15
Fixes: QTBUG-113152
Change-Id: Ifa08741cd621f2215eaf2f8fb848ecf6d5b9cf13
Reviewed-by: Alexey Edelev <alexey.edelev@qt.io>
Reviewed-by: Qt CI Bot <qt_ci_bot@qt-project.org>
(cherry picked from commit d2eaccfe7f613faa2083277d2d43c10c85053126)
---
 src/linguist/Qt5LinguistToolsMacros.cmake | 22 ++++++++++++++++------
 1 file changed, 16 insertions(+), 6 deletions(-)

diff --git a/src/linguist/Qt5LinguistToolsMacros.cmake b/src/linguist/Qt5LinguistToolsMacros.cmake
index 6a45e57be..0ccf59f3c 100644
--- a/src/linguist/Qt5LinguistToolsMacros.cmake
+++ b/src/linguist/Qt5LinguistToolsMacros.cmake
@@ -68,6 +68,7 @@ function(QT5_CREATE_TRANSLATION _qm_files)
     if(NOT EXISTS "${stamp_file_dir}")
         file(MAKE_DIRECTORY "${stamp_file_dir}")
     endif()
+    set(stamp_files "")
     foreach(_ts_file ${_my_tsfiles})
         get_filename_component(_ts_name ${_ts_file} NAME)
         if(_my_sources)
@@ -95,7 +96,14 @@ function(QT5_CREATE_TRANSLATION _qm_files)
 
           file(WRITE ${_ts_lst_file} "${_lst_file_srcs}")
         endif()
-        set(stamp_file "${stamp_file_dir}/${_ts_name}.stamp")
+        file(RELATIVE_PATH _ts_relative_path ${CMAKE_CURRENT_SOURCE_DIR} ${_ts_file})
+        string(REPLACE "../" "__/" _ts_relative_path "${_ts_relative_path}")
+        set(stamp_file "${stamp_file_dir}/${_ts_relative_path}.stamp")
+        list(APPEND stamp_files ${stamp_file})
+        get_filename_component(full_stamp_file_dir "${stamp_file}" DIRECTORY)
+        if(NOT EXISTS "${full_stamp_file_dir}")
+            file(MAKE_DIRECTORY "${full_stamp_file_dir}")
+        endif()
         add_custom_command(OUTPUT ${stamp_file}
             COMMAND ${Qt5_LUPDATE_EXECUTABLE}
             ARGS ${_lupdate_options} "@${_ts_lst_file}" -ts ${_ts_file}
@@ -103,7 +111,7 @@ function(QT5_CREATE_TRANSLATION _qm_files)
             DEPENDS ${_dependencies}
             VERBATIM)
     endforeach()
-    qt5_add_translation(${_qm_files} __QT_INTERNAL_DEPEND_ON_TIMESTAMP_FILE ${_my_tsfiles})
+    qt5_add_translation(${_qm_files} ${_my_tsfiles} __QT_INTERNAL_TIMESTAMP_FILES ${stamp_files})
     set(${_qm_files} ${${_qm_files}} PARENT_SCOPE)
 endfunction()
 
@@ -120,17 +128,17 @@ endif()
 
 
 function(QT5_ADD_TRANSLATION _qm_files)
-    set(options __QT_INTERNAL_DEPEND_ON_TIMESTAMP_FILE)
+    set(options)
     set(oneValueArgs)
-    set(multiValueArgs OPTIONS)
+    set(multiValueArgs OPTIONS __QT_INTERNAL_TIMESTAMP_FILES)
 
     cmake_parse_arguments(_LRELEASE "${options}" "${oneValueArgs}" "${multiValueArgs}" ${ARGN})
     set(_lrelease_files ${_LRELEASE_UNPARSED_ARGUMENTS})
 
+    set(idx 0)
     foreach(_current_FILE ${_lrelease_files})
         get_filename_component(_abs_FILE ${_current_FILE} ABSOLUTE)
         get_filename_component(qm ${_abs_FILE} NAME)
-        set(ts_stamp_file "${CMAKE_CURRENT_BINARY_DIR}/.lupdate/${qm}.stamp")
         # everything before the last dot has to be considered the file name (including other dots)
         string(REGEX REPLACE "\\.[^.]*$" "" FILE_NAME ${qm})
         get_source_file_property(output_location ${_abs_FILE} OUTPUT_LOCATION)
@@ -141,7 +149,9 @@ function(QT5_ADD_TRANSLATION _qm_files)
             set(qm "${CMAKE_CURRENT_BINARY_DIR}/${FILE_NAME}.qm")
         endif()
 
-        if(_LRELEASE___QT_INTERNAL_DEPEND_ON_TIMESTAMP_FILE)
+        if(_LRELEASE___QT_INTERNAL_TIMESTAMP_FILES)
+            list(GET _LRELEASE___QT_INTERNAL_TIMESTAMP_FILES ${idx} ts_stamp_file)
+            math(EXPR idx "${idx} + 1")
             set(qm_dep "${ts_stamp_file}")
         else()
             set(qm_dep "${_abs_FILE}")
-- 
2.45.1

